v3/lints/rfc/lint_dnsname_hyphen_in_sld.go

```diff
@@ -0,0 +1,58 @@
+package rfc
+
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"strings"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type DNSNameHyphenInSLD struct{}
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_rfc_dnsname_hyphen_in_sld",
+		Description:   "DNSName should not have a hyphen beginning or ending the SLD",
+		Citation:      "RFC5280: 4.2.1.6",
+		Source:        lint.RFC5280,
+		EffectiveDate: util.RFC5280Date,
+		Lint:          NewDNSNameHyphenInSLD,
+	})
+}
+
+func NewDNSNameHyphenInSLD() lint.LintInterface {
+	return &DNSNameHyphenInSLD{}
+}
+
+func (l *DNSNameHyphenInSLD) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubscriberCert(c) && util.DNSNamesExist(c)
+}
+
+func (l *DNSNameHyphenInSLD) Execute(c *x509.Certificate) *lint.LintResult {
+	parsedSANDNSNames := c.GetParsedDNSNames(false)
+	for i := range c.GetParsedDNSNames(false) {
+		if parsedSANDNSNames[i].ParseError != nil {
+			return &lint.LintResult{Status: lint.NA}
+		}
+		if strings.HasPrefix(parsedSANDNSNames[i].ParsedDomain.SLD, "-") ||
+			strings.HasSuffix(parsedSANDNSNames[i].ParsedDomain.SLD, "-") {
```

Comment:
It could make sense to document more of the citation why, but that's neither here nor there.

I believe the relevant language here is from RFC 1034 and relaxed by RFC 1123, namely, that the ABNF for a host in the "preferred name syntax" is

```text
<domain> ::= <subdomain> | " "
<subdomain> ::= <label> | <subdomain> "." <label>
<label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]
<ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
<let-dig-hyp> ::= <let-dig> | "-"
<let-dig> ::= <letter> | <digit>
```

which in 1034 terms, indicates a `<label>` begins with `<letter>` and ends with `let-dig`.  RFC 1123, Section 2.1, relaxes this, to become

```
<label> ::= <let-dig> [ [ <ldh-str> ] <let-dig> ]
```

by virtue of the following text:

```
      The syntax of a legal Internet host name was specified in RFC-952
      [DNS:4].  One aspect of host name syntax is hereby changed: the
      restriction on the first character is relaxed to allow either a
      letter or a digit.  Host software MUST support this more liberal
      syntax.
```

Note that 1123 references 952, which contained the following compatible-with-1034 ABNF

```
      <hname> ::= <name>*["."<name>]
      <name>  ::= <let>[*[<let-or-digit-or-hyphen>]<let-or-digit>]
```